### PR TITLE
Codejam canvas

### PR DESCRIPTION
**Pull request by o1vanova**
task - https://github.com/rolling-scopes-school/tasks/blob/master/tasks/stage-2/codejam-canvas/codejam-canvas.md
screen - https://www.flickr.com/gp/185210046@N03/Pnx3Nw
deploy - https://o1vanova.github.io/codejam-canvas/codejam-canvas/dist/index.html, https://tender-bell-67f493.netlify.com/
2019-09-27 / 27.09.2019
Total
Task ( 50 / 50 )
1. Basic scope (30):
+ Layout (menu, tools, canvas) - 10
+ Drawing color matrixes: (4x4 - 10, 32x32 - 10)
2. Extra scope (only if basic is implemented):
+ Show image on canvas 10
+ Dynamic canvas size switcher 10
Repository and commits requirements were implemented

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# dependencies
+/node_modules
+package-lock.json
```
Comment (magicsd) [inline .gitignore]: This file should be committed

File: codejam-canvas/src/index.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import './styles/site.scss';
+import {data as img4x4} from './assets/data/4x4';
+import { data as img32x32 } from './assets/data/32x32';
+
+const controls = document.getElementsByClassName('menu__control');
+const types = ['image', '4x4', '32x32'];
+
+let drawPicture = function () {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: Why are there so many 'let's? Do you plan to redeclare these variables?

```suggestion
const drawPicture = function () {
```

File: codejam-canvas/src/index.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    let img = new Image();
+    img.src = './assets/images/image.png';
+    let canvas = document.getElementById('canvas');
+    let ctx = canvas.getContext('2d');
+    ctx.fillStyle = "white";
+    ctx.fillRect(0, 0, 512, 512);
+
+    img.onload = function () {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]:
```javascript
img.addEventListener('load', () => {
  ctx.drawImage(img, (canvas.width / 2) - (img.width / 2), (canvas.height / 2) - (img.height / 2), img.width, img.height);
})
```

File: codejam-canvas/src/index.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        ctx.drawImage(img, (canvas.width / 2) - (img.width / 2), (canvas.height / 2) - (img.height / 2), img.width, img.height);
+    };
+}
+
+drawPicture();
+
+let drawCanvas = function (arr, divider, isRgb) {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: I'd prefer arrow function

File: codejam-canvas/src/index.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    let canvas = document.getElementById('canvas');
+    let ctx = canvas.getContext('2d');
+    //ctx.putImageData();
```
Comment (magicsd) [inline codejam-canvas/src/index.js]:
```suggestion
```

File: codejam-canvas/src/index.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    let x = 0;
+    let y = 0;
+    let n = 512/divider;
+    (function () {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: Why do you need IIFE here?

File: codejam-canvas/src/index.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                ctx.fillRect(x, y, n, n);
+                x += n;
+            }
+            x = 0;
+            y += n;
+        }
+    })();
+}
+
+for(let i = 0; i < controls.length; i++) {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]:
```javascript
controls.forEach(() => {})
```

File: codejam-canvas/src/index.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    controls[i].addEventListener('click', () => {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: I'd extract this anonymous function to named

File: codejam-canvas/src/styles/_canvas.scss [hunk 9]
```diff
@@ -0,0 +1,7 @@
+.canvas {
+    height: 512px;
+    width: 512px;
+    border: 1px solid $gray;
+    /*object-fit: contain;
+    object-position: center;*/
```
Comment (magicsd) [inline codejam-canvas/src/styles/_canvas.scss]: Shouldn't be here

File: codejam-canvas/src/styles/_header.scss [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    flex-direction: row;
+    justify-content: space-between;
+    align-items: center;
+    border-bottom: 2px solid $light-gray;
+    padding: 0 16px;
+
+    &__menu {
+        display: flex;
+        flex-direction: row;
+
```
Comment (magicsd) [inline codejam-canvas/src/styles/_header.scss]:
```suggestion
```

File: codejam-canvas/src/styles/_header.scss [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    &__title {
+        margin-left:24px;
```
Comment (magicsd) [inline codejam-canvas/src/styles/_header.scss]:
```suggestion
        margin-left: 24px;
```

File: codejam-canvas/src/styles/_variables.scss [hunk 12]
```diff
@@ -0,0 +1,13 @@
+//colors
+$white: #FFFFFF;
```
Comment (magicsd) [inline codejam-canvas/src/styles/_variables.scss]:
```suggestion
$white: #fff;
```

Comment (o1vanova) [reply codejam-canvas/src/styles/_variables.scss]: Is it faster because it hasn't opacity?

Comment (magicsd) [reply codejam-canvas/src/styles/_variables.scss]: It is nicer and in the same style 😊

File: codejam-canvas/src/index.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import './styles/site.scss';
+import {data as img4x4} from './assets/data/4x4';
+import { data as img32x32 } from './assets/data/32x32';
+
+const controls = document.querySelectorAll('.menu__control');
+const types = ['image', '4x4', '32x32'];
+
+const drawPicture = () => {
+    let img = new Image();
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: Please change all the `let`s to `const` if they aren't suppose to be redeclared. Here and everywhere.

File: codejam-canvas/src/index.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    img.src = './assets/images/image.png';
+    let canvas = document.getElementById('canvas');
+    let ctx = canvas.getContext('2d');
+    let zoom = 2;
+    ctx.fillStyle = "white";
+    ctx.fillRect(0, 0, 512, 512);
+
+    img.addEventListener('load', () => {
+        ctx.drawImage(img, (canvas.width / 2) - (img.width * zoom / 2), (canvas.height / 2) - (img.height * zoom / 2), img.width * zoom, img.height * zoom);
+    });
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: To make it clear what's happening here, I'd extract this callback to a named variable.
```javascript
const doSpecificStuff = () => {
  ctx.drawImage(img, (canvas.width / 2) - (img.width * zoom / 2), (canvas.height / 2) - (img.height * zoom / 2), img.width * zoom, img.height * zoom);
}

img.addEventListener('load', doSpecificStuff);
```

File: codejam-canvas/src/index.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    let canvas = document.getElementById('canvas');
+    let ctx = canvas.getContext('2d');
+    let x = 0;
+    let y = 0;
+    let n = 512/divider;
+
+    for (let i = 0; i < divider; i++) {
+        for (let j = 0; j < divider; j++) {
+            ctx.fillStyle = isRgb ? `rgba(${arr[j][i]})` : `#${arr[j][i]}`;
+            ctx.fillRect(x, y, n, n);
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: It's not clear what these `n`s are stand for. I guess they are `width` and `height`, so something like `rectSideLength` would be better.

File: codejam-canvas/src/index.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+const changeControl = (index) => {
+    for(let j = 0; j < controls.length; j++) {
+        if(j !== index) {
+            controls[j].classList.remove('menu-switcher__item--selected');
+        } else if (!controls[j].classList.contains('menu-switcher__item--selected')) {
+            controls[j].classList.add('menu-switcher__item--selected');
+        }
+    }
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: This kind seems more readable, especially when you name `index` and `cindex` in a clear way. Too much magic.
```javascript
const selector = 'menu-switcher__item--selected'

controls.forEach((control, cIndex) => {
  if (cIndex !== index) {
    control.classList.remove(selector)
  } else if (!control.classList.contains(selector)) {
    control.classList.add(selector);
  }
})
```

File: codejam-canvas/src/index.js [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    let zoom = 2;
+    ctx.fillStyle = "white";
+    ctx.fillRect(0, 0, 512, 512);
+
+    img.addEventListener('load', () => {
+        ctx.drawImage(img, (canvas.width / 2) - (img.width * zoom / 2), (canvas.height / 2) - (img.height * zoom / 2), img.width * zoom, img.height * zoom);
+    });
+}
+
+const drawCanvas = (arr, divider, isRgb) => {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: What is `arr`? Can name be a bit more clear? 😉

File: codejam-canvas/src/index.js [hunk 18]
```diff
@@ -1,67 +1,72 @@
...
+  const canvas = document.getElementById('canvas');
+  const ctx = canvas.getContext('2d');
+  const zoom = 2;
+  ctx.fillStyle = "white";
+  ctx.fillRect(0, 0, 512, 512);
 
-    img.addEventListener('load', () => {
-        ctx.drawImage(img, (canvas.width / 2) - (img.width * zoom / 2), (canvas.height / 2) - (img.height * zoom / 2), img.width * zoom, img.height * zoom);
-    });
+  const doSpecificStuff = () => {
```
Comment (magicsd) [inline codejam-canvas/src/index.js]: 🤣 I mean name it as the action it performs 😊

**Pull request merged**